### PR TITLE
fix(kafka): patch SerializingProducer and DeserializingConsumer without breaking their interfaces

### PR DESCRIPTION
**Pull request by emmettbutler**
This pull request fixes https://github.com/DataDog/dd-trace-py/issues/7174 by patching the `SerializingProducer` and `DeserializingConsumer` classes with classes that implement the appropriate functionality, rather than using `Consumer` and `Producer`. It also makes a few small changes to the Kafka integration for readability and the removal of Py2 support. 

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)
- [x] If this PR touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

### CODE REVIEW

File: tests/contrib/kafka/test_kafka.py [hunk 1]
```diff
@@ -129,18 +129,21 @@ def non_auto_commit_consumer(tracer, kafka_topic):
 
 @pytest.fixture
 def serializing_producer(tracer):
-    _producer = confluent_kafka.Producer({"bootstrap.servers": BOOTSTRAP_SERVERS})
+    _producer = confluent_kafka.SerializingProducer(
+        {"bootstrap.servers": BOOTSTRAP_SERVERS, "value.serializer": lambda x, y: x}
```
Comment (ZStriker19) [inline tests/contrib/kafka/test_kafka.py:133]: What's value.serializer? It looks like a lambda that just returns the first value it's called with?

File: ddtrace/internal/datastreams/kafka.py [hunk 2]
```diff
@@ -13,7 +13,7 @@
 INT_TYPES = (int,)
 
 
-def dsm_kafka_message_produce(instance, args, kwargs):
+def dsm_kafka_message_produce(instance, args, kwargs, is_serializing):
```
Comment (ZStriker19) [inline ddtrace/internal/datastreams/kafka.py:16]: When is this called with `is_serializing`?

Comment (ZStriker19) [reply ddtrace/internal/datastreams/kafka.py:16]: ah patch.py 142, got it! I wonder if there's an easier way to show these two things are connected or if that's something we'd want?

Comment (ZStriker19) [reply ddtrace/internal/datastreams/kafka.py:16]: Actually I guess it's pretty simple to search for `kafka.produce.start`.

File: tests/contrib/kafka/test_kafka.py (hunk 1 above)
Comment (emmettbutler) [reply tests/contrib/kafka/test_kafka.py:133]: That's right. It's a function that gets called with each message as an argument and that returns the message in a serialized form. https://docs.confluent.io/platform/current/clients/app-development.html

File: ddtrace/internal/datastreams/kafka.py (hunk 2 above)
Comment (emmettbutler) [reply ddtrace/internal/datastreams/kafka.py:16]: I've asked myself the same question in other contexts recently and at the moment I agree with you that Ctrl+F is good enough.

File: tests/contrib/kafka/test_kafka.py (hunk 1 above)
Comment (ZStriker19) [reply tests/contrib/kafka/test_kafka.py:135]: Ahhh got it, thanks!

**Review by ZStriker19 — DISMISSED** [review]

**Review by majorgreys — APPROVED** [review]

**Review by mabdinur — DISMISSED** [review]

**Comment by mabdinur** [comment]
Should we also backport this fix to 1.20?

**Review by ZStriker19 — APPROVED** [review]

**Pull request merged**